pubtools/pulplib/_impl/model/maintenance.py

```diff
@@ -0,0 +1,147 @@
+import logging
+import datetime
+import json
+import re
+
+import jsonschema
+
+from pubtools.pulplib._impl import compat_attr as attr
+from ..schema import load_schema
+from .common import InvalidDataException
+
+
+LOG = logging.getLogger("pubtools.pulplib")
+
+
+def _iso_time_now():
+    return datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")
+
+
+@attr.s(kw_only=True)
+class MaintainedRepo(object):
+    """Details about the maintenance"""
+
+    message = attr.ib(default=None, type=str)
+    """Why this repository is in maintenance"""
+    owner = attr.ib(default=None, type=str)
+    """Who set this repository in maintenance mode"""
+    started = attr.ib(default=None, type=str)
+    """When did the maintenance start"""
+
+
+@attr.s(kw_only=True)
+class MaintenanceReport(object):
+    """Represent the maintenance status of pulp"""
+
+    _SCHEMA = load_schema("maintenance")
+
+    last_updated = attr.ib(default=_iso_time_now(), type=str)
+    """:class:`~datetime.datetime` in UTC when this report was last updated,
+    if it's the first time the report is created, current time is used."""
+
+    last_updated_by = attr.ib(default="Content Delivery", type=str)
+    """Person/party who updated the report last time"""
+
+    repos = attr.ib(default={}, type=dict)
+    """A dictionary contains repo_id: :class:`MaintainedRepo` pairs, indicates
+    which repositories are in maintenance mode and details.
+    If empty, then it means no repositories is in maintenance mode.
+    """
+
+    @classmethod
+    def from_data(cls, data):
+        """Create a new report with a dictionary"""
+        try:
+            jsonschema.validate(instance=data, schema=cls._SCHEMA)
+        except jsonschema.exceptions.ValidationError as error:
+            LOG.exception("%s.from_data invoked with invalid data", cls.__name__)
+            raise InvalidDataException(str(error))
+
+        maintenance = cls(
+            last_updated=data["last_updated"],
+            last_updated_by=data["last_updated_by"],
+            repos={},
+        )
+        for repo_id, details in data["repos"].items():
+            maintenance.repos[repo_id] = MaintainedRepo(**details)
+
+        return maintenance
+
+    def json(self):
```

Comment:
I suggest not making this public API (or do you have a specific case in mind where something needs this?)

Of course, you need to be able to convert it to a raw dict locally, but we should discourage any other code except for this library making use of that raw format.

Reply:
Hm, I was just thinking about providing an alternative format of data to the user, if it's discouraged, then I guess this should be made private.